merkle-tree/src/merkle_tree.rs

```diff
@@ -139,6 +139,85 @@ impl MerkleTree {
         self.nodes.iter().last()
     }
 
+    fn append_nodes(nodes: &mut [Hash], mut leaf_count: usize, leaves: Vec<Hash>) -> usize {
+        for mut leaf in leaves {
+            let mut layer: usize = 0;
+            leaf_count += 1;
+            let mut cursor = leaf_count;
+            while (cursor & 1) == 0 {
+                let arg1 = &leaf;
+                let arg2 = &nodes[layer];
+                //todo: verify and implement jump's fancy avx implementation
+                leaf = hash_intermediate!(arg2, arg1);
+                layer += 1;
+                cursor >>= 1;
+            }
+
+            nodes[layer] = leaf;
+        }
+        leaf_count
+    }
+
+    fn private_depth(leaf_count: usize) -> usize {
+        if leaf_count <= 1 {
+            leaf_count
+        } else {
+            (63 - (leaf_count - 1).leading_zeros()) as usize + 2
+        }
+    }
+
+    fn commit_finish(nodes: &mut [Hash], leaf_count: usize) -> Hash {
+        let root_idx: usize = Self::private_depth(leaf_count) - 1;
+
+        if !leaf_count.is_power_of_two() {
+            let mut layer = leaf_count.trailing_zeros() as usize;
+            let mut layer_count = leaf_count >> layer;
+            let mut tmp = nodes[layer];
+
+            while layer_count > 1 {
+                tmp = if (layer_count & 1) != 0 {
+                    let arg1 = &tmp;
+                    let arg2 = &tmp;
+                    hash_intermediate!(arg2, arg1)
+                } else {
+                    let arg1 = &tmp;
+                    let arg2 = &nodes[layer];
+                    hash_intermediate!(arg2, arg1)
+                };
+                layer += 1;
+                layer_count = (layer_count + 1) >> 1;
+            }
+            nodes[root_idx] = tmp;
+        }
+        nodes[root_idx]
+    }
+
+    // Saves memory when one only needs the root
+    // of a merkle tree by computing the root of the merkle
+    // tree for the given data without building up a merkle tree
+    #[allow(clippy::uninit_vec)]
+    pub fn merkle_root<T: AsRef<[u8]>>(items: &[T]) -> Option<Hash> {
+        if items.is_empty() {
+            return None;
+        }
+        let mut nodes: Vec<Hash> = Vec::with_capacity(63);
+        unsafe {
+            nodes.set_len(63);
+        }
+
+        let hashes = items
+            .iter()
+            .map(|item| {
+                let item = item.as_ref();
+                hash_leaf!(item)
+            })
+            .collect::<Vec<_>>();
+        let mut leaf_count = 0;
+        leaf_count = Self::append_nodes(&mut nodes, leaf_count, hashes);
```

Comment:
Isn't this `leaf_count` here just `items.len()`?

Also `append_nodes` is meant to be used in a streaming manner, where we don't have all leaves at once.
If the public method is only being used with all leaves available, then I believe this code could be simpler.
e.g. don't need to pass around `leaf_count` anymore.